R/plotAbundanceDensity.R

```diff
@@ -0,0 +1,259 @@
+        density_data <- density_data_list$density_data
+        colour_by <- density_data_list$colour_by
+        
+        # Gets the plot from plotter
+        plot_out <- .density_plotter(density_data = density_data, 
+                                     layout = layout,
+                                     xlab = abund_values,
+                                     ylab = "Taxa",
+                                     colour_by = colour_by,
+                                     ...)
+        return(plot_out)
+    }
+)
+
+################################ HELP FUNCTIONS ################################
+
+.incorporate_density_data <- function(object, abund_values, n, colour_by){
+    # Gets the assay
+    mat <- assay(object, abund_values)
+    # Gets the most abundant taxa
+    top_taxa <- getTopTaxa(object, n)
+    # Subsets abundance table  by taking taxa of highest abundance
+    mat <- mat[top_taxa, , drop=FALSE]
+    # melt the data
+    density_data <- t(mat) %>%
+        as.data.frame() %>%
+        rownames_to_column("Sample") 
+    # Gets coloring information if 'colour_by' is not NULL
+    if (!is.null(colour_by)) {
+        # Gets information from colData
+        colour_out <- retrieveCellInfo(object, colour_by)
+        # Mirrors back the variable name, if a partial match was used
+        colour_by <- colour_out$name
+        # Adds information to the data frame that includes all density data
+        density_data$colour_by <- colour_out$value
+    }
+    cols <- c("Sample","colour_by")[c("Sample","colour_by") %in% colnames(density_data)]
+    density_data <- density_data %>%
+        pivot_longer(cols = !cols, names_to = "Y", values_to = "X")
+    # Converts taxa to factor. Order of levels is the opposite than in 'top_taxa'
+    # so that taxa with highest abundance is on top
+    density_data$Y <- factor( density_data$Y, rev(top_taxa) )
+    return(list(density_data = density_data, 
+                colour_by = colour_by))
+}
+
+.density_plotter <- function(density_data, 
+                             layout,
+                             xlab = NULL,
+                             ylab = NULL,
+                             colour_by = NULL,
+                             point_alpha = 0.6,
+                             point_shape = 21,
+                             point_size = 2,
+                             add_legend = TRUE,
+                             flipped = FALSE,
+                             add_x_text = TRUE,
+                             log_scale_x_axis = FALSE){
+    # start plotting
+    # Density plot needs different kind of structure
+    if (layout == "density"){
+        # # Reorders the levels to reverse order so that the taxa with highest abundance
+        # is on the top
+        density_data$Y <- factor(density_data$Y , levels = rev(levels(density_data$Y)) )
+        plot_out <- ggplot(density_data, aes_string(x="X", colour = "Y")) +
+            xlab(xlab) +
+            ylab(ylab)
```

Comment:
I don't think you should colour by "Y". The colour can then not be changed by the user. I think you should keep it as `aes_string(x="X", colour = "Y")` and supplement a specific different aesthetic in the function call

Reply:
I have been trying to do that way. That is why I chose to do density plots with facets, because I couldn't find a solution to arrange multiple density plots to the same picture

Reply:
I will have a look

Reply:
Have a look at the changes

Reply:
Now it looks very nice. `shape_by` and `size_by` wasn't working due to small bug, so I fixed them. Also I added some information to man page